erc20-bridge-token/contracts/BridgeTokenFactory.sol

```diff
@@ -0,0 +1,152 @@
+import "@openzeppelin/contracts-upgradeable/utils/math/SafeMathUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+
+import "rainbow-bridge-sol/nearprover/contracts/INearProver.sol";
+import "rainbow-bridge-sol/nearprover/contracts/ProofDecoder.sol";
+import "rainbow-bridge-sol/nearbridge/contracts/NearDecoder.sol";
+import "rainbow-bridge-sol/nearbridge/contracts/Borsh.sol";
+
+import "./Locker.sol";
+import "./BridgeToken.sol";
+import "./BridgeTokenProxy.sol";
+
+contract BridgeTokenFactory is Locker, AccessControlUpgradeable, PausableUpgradeable{
+
+    using Borsh for Borsh.Data;
+    using SafeMathUpgradeable for uint256;
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+    
+    mapping(address => string) private _ethToNearToken;
+    mapping(string => address) private _nearToEthToken;
+    mapping(address => bool) private _isBridgeToken;
+
+    bytes32 public constant PAUSE_ROLE = keccak256("PAUSE_ROLE");
+    // Event when funds are withdrawn from Ethereum back to NEAR.
+    event Withdraw (
+        string token,
+        address indexed sender,
+        uint256 amount,
+        string recipient
+    );
+
+    event Deposit (
+        uint256 amount,
+        address recipient
+    );
+
+    event SetMetadata (
+        address token,
+        string name,
+        string symbol,
+        uint8 decimals
+    );
+
+    struct LockResult {
+        string token;
+        uint128 amount;
+        address recipient;
+    }
+
+    struct MetadataResult {
+        string token;
+        string name;
+        string symbol;
+        uint8 decimals;
+        uint64 blockHeight;
+    }
+
+
+    // BridgeTokenFactory is linked to the bridge token factory on NEAR side.
+    // It also links to the prover that it uses to unlock the tokens.
+    function initialize(bytes memory _nearTokenFactory, INearProver _prover,  uint64 _minBlockAcceptanceHeight)
+     public initializer{
+        __AccessControl_init();
+        __Pausable_init_unchained();
+        __Locker_init(_nearTokenFactory, _prover, _minBlockAcceptanceHeight);
+        _setupRole(DEFAULT_ADMIN_ROLE, _msgSender()); 
+        _setupRole(PAUSE_ROLE, _msgSender()); 
+    
+    }
+
+    function isBridgeToken(address token) external view returns (bool) {
+        return _isBridgeToken[token];
+    }
+
+    function ethToNearToken(address token) external view returns (string memory) {
+        require(_isBridgeToken[token], "ERR_NOT_BRIDGE_TOKEN");
+        return _ethToNearToken[token];
+    }
+
+    function nearToEthToken(string calldata nearTokenId) external view returns (address) {
+        require(_isBridgeToken[_nearToEthToken[nearTokenId]], "ERR_NOT_BRIDGE_TOKEN");
+        return _nearToEthToken[nearTokenId];
+    }
+
+    function newBridgeToken(string calldata nearTokenId) external returns (BridgeTokenProxy) {
```

Comment:
Please add scripts for upgrading erc20 token (for this we need to add upgradeBridgeToken because as I see the factory is admin for the proxy)

Reply:
ac626525272f734706fcc2a2e291cc5e7b20f0ff